File: State_Pattern/cpp/State_Pattern.cc
#include <iostream>
using namespace std;

class Vote {
public:
    virtual void DoAction() = 0;
};

class UnVote : public Vote {
public:
    void DoAction() {
        cout << "vote success" << endl;
    }
};

class Voted : public Vote {
public:
    void DoAction() {
        cout << "already voted" << endl;
    }
};

class Votor {
private:
    bool state;
    Vote* v;
public:
    Votor():state(false),v(new UnVote){}
public:
    void DoVote() {
        if (state) {
            v = new Voted;
        }
        v->DoAction();
        state = true;
    }
};

void test_state() {
    Votor v;
    v.DoVote();
    v.DoVote();
}

int main() {
    test_state();
    return 0;
}

File: State_Pattern/go/State_Pattern.go
package State_Pattern

import "fmt"

type Vote interface {
	DoAction()
}

type Voted struct{}

func (*Voted) DoAction() {
	fmt.Println("already voted")
}

type UnVote struct{}

func (*UnVote) DoAction() {
	fmt.Println("vote success")
}

type Votor struct {
	isDone bool
	doDove Vote
}

func (this *Votor) DoVote() {
	if !this.isDone {
		this.doDove = &UnVote{}
	}
	this.doDove.DoAction()
	this.isDone = true
	this.doDove = &Voted{}
}
